treyholt/Suzie
# Handling Chrome LifeCycle Events with Zara.com and parsing cart items of Zara

See the Technical Hurdles bit added below (but also in Wiki) for a further breakdown of the problems faced with this commit. 

This PR accomplishes a few things. First and foremost it parses the HTML of a Zara.com cart to extract important info about the items in the cart. Currently we can extract the name, color and size. We have no use for it yet, but that will come. 

Then it also handles the edge case of some oddities around Chrome life cycle events. 

**To Test**
Please test this PR, as I want to make sure it works on other people's devices the same way it does on mine. 
When you've properly refreshed this code into your Suzie extension on your browser, go to Zara.com 
Add a few items to your cart, then go to the cart.
Check the console 
you should see a printed array of information about those items: color, name and size. 


### Chrome Web Page Life Cycle Struggles

Currently our content scripts run when chrome decides a webpage document has fully loaded from a proper url change. 

We specify this in the manifest.json file on line 19 like this: 
`"run_at": "document_end"`

This means that at the time our javascript is running in app.js, we know what is laid out and so we can inspect and manipulate all of the available HTML. 

This is the case almost every time it seems a URL changes, certainly when the hostname changes. But I've noticed that when navigating from say an item detail page on Zara to the shopping cart, this navigation changes only the subdirectory of the URL. This evidently doesn't change the URL enough to trigger our content scripts to run when the document is loading. 

To handle this oddity I had to look into how we can manage and understand the lifecycle of the web page we are working with in our content scripts. It turns out, using Chrome's 'Tabs' events we can recognize when a URL changes on any tab. From there we can send a message through chrome, to hit a callback listener in our content script that can again manipulate the document tree (parse important info etc). The problem is, the recognition that a url has changed happens before the DOM is actually laid out, meaning at that point, we can't actually do anything meaningful in the script pertaining to the document tree. 

This leads to the current solution we have--which I am not particularly proud of. 
Since our script runs whenever the page refreshes, in that callback listener, instead of trying to handle the document tree, we can just refresh the page, triggering our content scripts again when they can properly handle and parse the HTML.
